sdk/core/Azure.Core/src/ResourceIdentifier.cs

```diff
@@ -0,0 +1,429 @@
+            var id = Create(resourceId);
+            Init(id.Parent, id.ResourceType, id.Name);
+            ResourceType = id.ResourceType;
+            Name = id.Name;
+            Parent = id.Parent;
+            IsChild = id.IsChild;
+            _stringValue = resourceId;
+        }
+
+        private ResourceIdentifier(ResourceIdentifier? parent, ResourceType resourceType, string resourceName, bool isChild)
+        {
+            Init(parent, resourceType, resourceName);
+            ResourceType = resourceType;
+            Name = resourceName;
+            Parent = parent ?? Root;
+            IsChild = isChild;
+            if (parent is null)
+                _stringValue = RootStringValue;
+        }
+
+        private ResourceIdentifier(ResourceIdentifier parent, string resourceTypeName, string resourceName)
+            : this(parent, ChooseResourceType(resourceTypeName, parent), resourceName, true)
+        {
+        }
+
+        private ResourceIdentifier(ResourceIdentifier parent, string providerNamespace, string resourceTypeName, string resourceName)
+            : this(parent, new ResourceType(providerNamespace, resourceTypeName), resourceName, false)
+        {
+        }
+
+        private void Init(ResourceIdentifier? parent, ResourceType resourceType, string resourceName)
+        {
+            if (parent is not null)
+            {
+                Provider = parent.Provider;
+                SubscriptionId = parent.SubscriptionId;
+                Location = parent.Location;
+                ResourceGroupName = parent.ResourceGroupName;
+            }
+
+            if (resourceType == SubscriptionResourceType)
+            {
+                if (!Guid.TryParse(resourceName, out _))
+                    throw new ArgumentOutOfRangeException(nameof(resourceName), $"The GUID for subscription is invalid {resourceName}.");
+                SubscriptionId = resourceName;
+            }
+
+            if (resourceType.GetLastType() == LocationsKey)
+                Location = resourceName;
+
+            if (resourceType == ResourceGroupResourceType)
+                ResourceGroupName = resourceName;
+
+            if (resourceType == ProviderResourceType)
+                Provider = resourceName;
+        }
+
+        private static ResourceType ChooseResourceType(string resourceTypeName, ResourceIdentifier parent)
+        {
+            if (resourceTypeName.Equals(ResourceGroupsLowerKey, StringComparison.OrdinalIgnoreCase))
+                return ResourceGroupResourceType;
+
+            //subscriptions' type is Microsoft.Resources/subscriptions only when its parent is Tenant
+            if (resourceTypeName.Equals(SubscriptionsKey, StringComparison.OrdinalIgnoreCase) && parent.ResourceType == TenantResourceType)
+                return SubscriptionResourceType;
+
+            return parent.ResourceType.AppendChild(resourceTypeName);
+        }
+
+        private static ResourceIdentifier Create(string resourceId)
```

Comment:
Is this better called Parse?  We're doing validation and then making a recursive call.  Each recursive call creates a node in the tree.